### PR TITLE
Add definition of StreakRendererClass and StreakLineClass.

### PR DESCRIPTION
**Pull request by jonwil**
Hook StreakLineClass::Prepare_LOD so that we can make
PredictiveLODOptimizerClass go away

### CODE REVIEW

File: src/w3d/renderer/streak.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    virtual int Get_LOD_Count() const override;
+
+    void Reset_Line();
+
+    void Set_Locs(unsigned int count, Vector3 *locs);
+    void Set_Widths(unsigned int count, float *widths);
+    void Set_Colors(unsigned int count, Vector4 *colors);
+    void Set_Locs_Widths_Colors(unsigned int count, Vector3 *locs, float *widths, Vector4 *colors, unsigned int *unk);
+
+    int Get_Num_Points();
```
Comment (xezon) [inline src/w3d/renderer/streak.h:53]: Can be `const` ?

File: src/w3d/renderer/streak.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    void Set_Point_Location(unsigned int point_idx, const Vector3 &location);
+    void Get_Point_Location(unsigned int point_idx, Vector3 &loc);
```
Comment (xezon) [inline src/w3d/renderer/streak.h:55]: Can be `const` ?

File: src/w3d/renderer/hlod.cpp [hunk 3]
```diff
@@ -1535,15 +1534,12 @@ void HLodClass::Prepare_LOD(CameraClass &camera)
     if (Is_Not_Hidden_At_All()) {
         float sz = Get_Screen_Size(camera);
 
-        if (m_lodCount <= 1) {
-            PredictiveLODOptimizerClass::Add_Cost(Get_Cost());
-        } else {
+        if (m_lodCount > 1) {
```
Comment (xezon) [inline src/w3d/renderer/hlod.cpp:1537]: Should there be a comment that this now deviates from original?

File: src/w3d/renderer/part_buf.cpp [hunk 4]
```diff
@@ -1190,8 +1189,6 @@ void ParticleBufferClass::Prepare_LOD(CameraClass &camera)
     if (Get_LOD_Level() < minlod) {
         Set_LOD_Level(minlod);
     }
-
-    PredictiveLODOptimizerClass::Add_Object(this);
```
Comment (xezon) [inline src/w3d/renderer/part_buf.cpp:1194]: Should there be a comment that this now deviates from original?

File: src/w3d/renderer/segline.cpp [hunk 5]
```diff
@@ -339,12 +338,6 @@ void SegmentedLineClass::Prepare_LOD(CameraClass &camera)
     }
 
     m_lineRenderer.Set_Current_Subdivision_Level(level);
-
-    if (m_maxSubdivisionLevels) {
```
Comment (xezon) [inline src/w3d/renderer/segline.cpp:343]: Should there be a comment that this now deviates from original?

**Review by xezon — APPROVED** [review]
Logically looks ok. Not tested in game.

File: src/w3d/renderer/streak.h [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    void Set_Merge_Intersections(int onoff);
+    void Set_Freeze_Random(int onoff);
+    void Set_Disable_Sorting(int onoff);
+    void Set_End_Caps(int onoff);
+
+    void Render_Seg_Line(RenderInfoClass &rinfo);
+    void Render_Streak_Line(RenderInfoClass &rinfo);
+
+private:
+    int m_subdivisionLevels;
```
Comment (tomsons26) [inline src/w3d/renderer/streak.h]: m_maxSubdivisionLevels

File: src/w3d/renderer/streak.h [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    unsigned int *m_unk;
+    float m_screenSize;
```
Comment (tomsons26) [inline src/w3d/renderer/streak.h]: m_normalizedScreenArea

File: src/w3d/renderer/streak.h [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    SimpleDynVecClass<Vector3> m_pointLocations;
+    SimpleDynVecClass<Vector4> m_colors;
```
Comment (tomsons26) [inline src/w3d/renderer/streak.h]: m_pointColors
m_pointWidths

File: src/w3d/renderer/streak.h [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+    SimpleDynVecClass<float> m_widths;
+    SegLineRendererClass m_segLine;
```
Comment (tomsons26) [inline src/w3d/renderer/streak.h]: m_lineRenderer
m_streakRenderer

File: src/w3d/renderer/streakrender.h [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    VertexFormatXYZUV1 *Get_Vertex_Buffer(unsigned int vertex_count);
+
+    TextureMapMode Get_Texture_Mapping_Mode() const;
+    void Set_Texture_Mapping_Mode(TextureMapMode mode);
+
+    int Is_Freeze_Random() const { return m_bits & FREEZE_RANDOM; }
+    int Is_Merge_Intersections() const { return m_bits & MERGE_INTERSECTIONS; }
+    int Is_Sorting_Disabled() const { return m_bits & DISABLE_SORTING; }
+    void Set_Color(const Vector3 &color) { m_color = color; }
+    void Set_Current_Subdivision_Level(unsigned int lv) { m_subdivisionLevel = lv; }
```
Comment (tomsons26) [inline src/w3d/renderer/streakrender.h]: probably move to cpp and make this the way it is in bfme2, with assert and setting to 0

File: src/w3d/renderer/streak.h (hunk 6 above)
Comment (jonwil) [reply src/w3d/renderer/streak.h]: fixed.

File: src/w3d/renderer/streak.h (hunk 7 above)
Comment (jonwil) [reply src/w3d/renderer/streak.h]: fixed.

File: src/w3d/renderer/streak.h (hunk 8 above)
Comment (jonwil) [reply src/w3d/renderer/streak.h]: fixed.

File: src/w3d/renderer/streak.h (hunk 9 above)
Comment (jonwil) [reply src/w3d/renderer/streak.h]: fixed.

File: src/w3d/renderer/streakrender.h (hunk 10 above)
Comment (jonwil) [reply src/w3d/renderer/streakrender.h]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**